### ISSUE TITLE
Editor's scripting proposal

### ISSUE THREAD

**Issue by ekharkunov** [defold/defold#8504]
_Is it a private platform specific feature? (i.e. would it break your NDA?)<br>_
No

**Is your feature request related to a problem? Please describe (REQUIRED):**
Some of project/extensions should make assets generation/modification. Introduce new API which add possibility to generate/modify project's assets.

**Use cases:**
1. Editor scripts which update atlas with all images from folder
You add/remove image - with editor's script you list all images from folder and recreate atlas asset with new API.

2. Generate "spawner" objects
For example, you have some king of strategy game. In your game you can create new type of units manually and after that you should update game object that used to spawn new unit type in runtime. I mean you should add new factory/collectionfactory component to "spawner" object.

3. Extend previous case - generate game objects/collections from balance table
For example, you have some Excel/Google sheets tables that can be exported to csv and stored inside repository. During build phase you need to generate game objects/collections/other assets related to balance table. It helps to make assets to be consistent with balance tables.

4. Generate collection for level
For example, you used Photoshop to combine different sprites and create a game level. Use tool like PsdSplit to create a bunch of sprites and xml description with sprite position. After that use editor's script to create set of game objects, atlas with all sprites and level collection where you place game objects in the right place according to xml description.

**Describe the solution you'd like (REQUIRED):**
Expose more API in editor's script for:
1. Creating all type of assets: atlas, game object, collection, etc.
2. Modify assets: 
2a. add/remove components
2b. modify atlas content/settings
2c. modify game objects hierarchy inside collection
2d. change component properties
3. Iterating over nested data
4. Getting component/game object/piece of data via id (see example below)

**Describe alternatives you've considered (REQUIRED):**
You can use any external scripted or compiled language to create external tool to generate/modifying assets. For now workflow can be looked like:
1. Download defoldsdk.zip according to pinned version
2. Unzip archive
3. Generate code for certain language from proto definitions (if it scripting language like Python) (optional)
4. Run script or compiled program to make asset generation/modification

**Additional context (OPTIONAL):**
Some examples how it can be looks like:
1. Group game object as children of newly created empty game object
```
local work_collection = assets.get_collection("/main/test/main.collection")
local new_parent = assets.collection.create_game_object("new_parent_id")
for id, game_object in pairs(work_collection.get_game_objects()) do
    if string.find(id, "tile_") then
        assets.collection.set_parent(game_object, new_parent, {}) --- some additional options
    end
end
assets.save(work_collection, assets.get_path(work_collection)) --- can be assets.save(work_collection, "<path_to_new_asset_location")
editor.complete_transaction() --- resync editor's state
```

2. Add all images from folder to atlas
```
local atlas_path = "/main/atlases/rgb/level1/tiles.atlas"
local images = editor.list_directory("/assets/level1/tiles/")
editor.remove(atlas_path)
local atlas_asset = assets.create_atlas()
atlas.margin = 2
atlas.padding = 1
atlas.extrude_border = 2
for _, image_path in ipairs(images) do
    assets.atlas.add_image(atlas, image_path, assets.atlas.TRIM_MODE.VERTICES_7) -- last parameter can be optional
end
assets.save(atlas, atlas_path)
editor.complete_transaction()
```

**Comment by wistpotion** [comment]
This might not be entirely what you have in mind, but currently you can [add menu items](https://defold.com/manuals/editor-scripts/#commands). This is what monarch does with it's scene creation. Monarch is actually doing some file modification with it's menu item, it might be worth taking a peek into [this file](https://github.com/britzl/monarch/blob/master/monarch/editor-script/make_monarch.editor_script).

So you can modify every asset (though the action is not undoable) in your project (which includes generating stuff for collections). It just takes a lot of wiggling around with the asset files. So maybe a nice lua wrapper for editing asset data could be a way of achieving what you want to do?

Then the flow would be something along the lines of 
defold menu item ->
read the asset file ->
modify whatever you want with it (remember that this works with collections as well!!) ->
save the modified asset file ->
profit

**Comment by ekharkunov** [comment]
Yes, every assets modification can be done via plain lua because assets store tin text form of protobuf. It can be done like make string template with placeholders and replacing it (something like monarch do https://github.com/britzl/monarch/blob/master/monarch/editor-script/make_monarch.editor_script#L117). But it become harder to do when you operate with complex structure. For example, storing component description in `data` field (like https://github.com/britzl/monarch/blob/master/example/advanced/menu.collection#L87)

Those proposal is more about providing API to work with assets because under the hood every asset is simple protobuf structure. Also when you work with assets via API you don't depend on how asset stored. I mean any changes to protobuf structure doesn't break you script because you don't use string template.

**Comment by vlaaad** [comment]
Thank you for the suggestion @ekharkunov!

Me and @matgis discussed your feature request. We conclude that editor extensions should be expanded, and modifying components is a very useful extension point. There might be different approaches to providing an API for the functionality, e.g. using an imperative API, using transactions similar to how we have a `"set"` action, or using some blend of it as we are considering in #8427.

We also agree that working with resources as simple data structures (which the protobufs are) is useful, although some trade-offs need to be considered first. We store the data in the editor data graph in a relational way, so fine-grained transactions may propagate and update the related references. This means that while the approach of modifying the whole structure in a single step is simpler for the user, in some cases, it might lead to broken references. We still think this is very useful, so we might implement it in the future, but we generally want the main API to be more fine-grained.

Regarding building Defold resources from user data formats like Excel sheets: this is a slightly different functionality than the use cases 1 and 2. For such builders, we probably want to support them both in the editor and Bob. This idea was on the back of our minds for some time already, see #4503. We want this feature eventually, but we haven't decided on the exact implementation yet.

We will keep your suggestion in mind when deciding what to do next for the editor scripts!

**Comment by Jerakin** [comment]
Poking my head in here only to mention that one of the first editor-scripts I created was one that adds images to atlases. If you are looking for that functionality.

https://github.com/Jerakin/editor-script-atlas

And +1 to improving editor scripting in general!

**Comment by vlaaad** [comment]
I believe this is done!
